File: C++/GraphsCPP/IslandPerimeter.cpp
class Solution {
public:
    int count = 0;
    void dfs(vector<vector<int>>& grid, int i, int j, vector<vector<bool>>& visited){
        int n = grid.size();
        int m = grid[0].size();
        
        // Overflows or side is water then also perimeter increases by 1
        if(i < 0  ||  j < 0  ||  i >= n  ||  j >=m  ||  grid[i][j]==0){
            count++;
            return;
        }
        
        if(visited[i][j])
            return;
        
        visited[i][j] = true;
        
        dfs(grid, i+1, j, visited);
        dfs(grid, i-1, j, visited);
        dfs(grid, i, j+1, visited);
        dfs(grid, i, j-1, visited);
    }
    
    int islandPerimeter(vector<vector<int>>& grid) {
        count = 0;
        
        int n = grid.size();
        int m = grid[0].size();
        
        vector<vector<bool>> visited(n, vector<bool>(m, 0));
        
        for(int i = 0; i < n; i++){
            for(int j = 0; j < m; j++){
                if(!visited[i][j]  &&  grid[i][j] == 1){
                    dfs(grid, i, j, visited);
                    
                    //Only one island is present as stated in question hence break
                    break;
                }
            }
        }
        
        return count;
    }
};

File: JAVA/GraphsJava/IslandPerimeter.java
/*
You are given row x col grid representing a map where grid[i][j] = 1 represents land and grid[i][j] = 0 represents water.

Grid cells are connected horizontally/vertically (not diagonally). The grid is completely surrounded by water, and there is exactly one island (i.e., one or more connected land cells).

The island doesn't have "lakes", meaning the water inside isn't connected to the water around the island. One cell is a square with side length 1. The grid is rectangular, width and height don't exceed 100. Determine the perimeter of the island.

 

Example 1:


Input: grid = [[0,1,0,0],[1,1,1,0],[0,1,0,0],[1,1,0,0]]
Output: 16
Explanation: The perimeter is the 16 yellow stripes in the image above.
*/


class Solution {
    int count  = 0;
    public int islandPerimeter(int[][] grid) {
        count = 0;
        for(int i = 0 ;i<grid.length ;i++){
            for(int j = 0;j<grid[i].length;j++){
                if(grid[i][j] == 1){
                     dfs(grid,i,j);
                    //break as we got island
                     break;
                }
            }
        }
       return count; 
    }
    
    private void dfs(int[][]grid,int i,int j){
        //ovrflow will also add to count
        if(i<0 || i>=grid.length || j<0 || j>=grid[i].length || grid[i][j] == 0 ){
           count++;
          return;
        }
        //check if already visited
        if(grid[i][j] == -1)return;
        //mark visited
        grid[i][j] = -1;
        //for all the 4 directions
        dfs(grid,i+1,j);
        dfs(grid,i-1,j);
        dfs(grid,i,j+1);
        dfs(grid,i,j-1);
    }
}
